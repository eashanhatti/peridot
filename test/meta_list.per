axiom List: Function(MetaType) ~> MetaType;
axiom nil: Function(A: MetaType) ~> List(A);
axiom cons: Function(A: MetaType, A, List(A)) ~> List(A);

axiom Append:
  Function(
    A: MetaType,
    List(A),
    List(A),
    List(A)) ~> MetaType;
axiom append_nil:
  Forall A, Forall xs,
  Append(A, xs, nil(A), nil(A));
axiom append_cons:
  Forall A, Forall x, Forall ys, Forall zs, Forall ws,
  Implies(
    Append(A, ys, zs, ws),
    Append(A, cons(A, x, ys), zs, cons(A, x, ws)));

axiom Unit: MetaType;
axiom unit: Unit;

variable xs: List(Unit);
proof Append(Unit, nil(Unit), nil(Unit), xs);

variable ys: List(Unit);
proof Append(
  Unit,
  cons(Unit, unit, nil(Unit)),
  cons(Unit, unit, nil(Unit)),
  ys);