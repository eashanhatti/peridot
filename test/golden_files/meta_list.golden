((fromList
    [(UVGlobal 8,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 1})))))
              (Neutral _ (UniVar (LVGlobal 2000)))))),
     (UVGlobal 1000, Nothing),
     (UVGlobal 1001, Just (Rigid (TypeType Meta))),
     (UVGlobal 1002, Just (Rigid (TypeType Meta))),
     (UVGlobal 1003, Nothing),
     (UVGlobal 1004,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1005, Just (Rigid (TypeType Meta))),
     (UVGlobal 1006, Nothing),
     (UVGlobal 1007, Just (Rigid (TypeType Meta))),
     (UVGlobal 1008, Just (Rigid (TypeType Meta))),
     (UVGlobal 1009, Nothing),
     (UVGlobal 1010, Just (LocalVar Level {unLevel = 0})),
     (UVGlobal 1011, Just (Rigid (TypeType Meta))),
     (UVGlobal 1012, Nothing),
     (UVGlobal 1013,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1014, Just (Rigid (TypeType Meta))),
     (UVGlobal 1015, Nothing),
     (UVGlobal 1016,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1017, Just (Rigid (TypeType Meta))),
     (UVGlobal 1018, Nothing),
     (UVGlobal 1019,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1020, Just (Rigid (TypeType Meta))),
     (UVGlobal 1021, Nothing),
     (UVGlobal 1022, Just (Rigid (TypeType Meta))),
     (UVGlobal 1023, Just (Rigid (TypeType Meta))),
     (UVGlobal 1024, Nothing),
     (UVGlobal 1025, Just (LocalVar Level {unLevel = 0})),
     (UVGlobal 1026, Just (Rigid (TypeType Meta))),
     (UVGlobal 1027, Nothing),
     (UVGlobal 1028,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1029, Just (Rigid (TypeType Meta))),
     (UVGlobal 1030, Nothing),
     (UVGlobal 1031,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1032, Just (Rigid (TypeType Meta))),
     (UVGlobal 1033, Nothing),
     (UVGlobal 1034,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1035, Just (Rigid (TypeType Meta))),
     (UVGlobal 1036, Nothing),
     (UVGlobal 1037, Just (Rigid (TypeType Meta))),
     (UVGlobal 1038, Just (Rigid (TypeType Meta))),
     (UVGlobal 1039, Nothing),
     (UVGlobal 1040,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1041, Just (Rigid (TypeType Meta))),
     (UVGlobal 1042, Nothing),
     (UVGlobal 1043, Just (Rigid (TypeType Meta))),
     (UVGlobal 1044, Just (Rigid (TypeType Meta))),
     (UVGlobal 1045, Nothing),
     (UVGlobal 1046,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1047, Just (Rigid (TypeType Meta))),
     (UVGlobal 1048, Nothing),
     (UVGlobal 1049, Just (Rigid (TypeType Meta))),
     (UVGlobal 1050, Just (Rigid (TypeType Meta))),
     (UVGlobal 1051, Nothing),
     (UVGlobal 1052, Just (LocalVar Level {unLevel = 0})),
     (UVGlobal 1053, Just (Rigid (TypeType Meta))),
     (UVGlobal 1054, Nothing),
     (UVGlobal 1055,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1056, Just (Rigid (TypeType Meta))),
     (UVGlobal 1057, Nothing),
     (UVGlobal 1058,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1059, Just (Rigid (TypeType Meta))),
     (UVGlobal 1060, Nothing),
     (UVGlobal 1061,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0})))))
              (LocalVar Level {unLevel = 0})))),
     (UVGlobal 1062, Just (Rigid (TypeType Meta))),
     (LVGlobal 2000,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Meta (Id {unId = 6})))))),
     (LVGlobal 2001,
      Just
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (GlobalVar (Rigid (NameIntro Meta (Id {unId = 1})))))
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Meta (Id {unId = 6}))))))))],
  fromList
    [(UVGlobal 1000, UVGlobal 1001),
     (UVGlobal 1001, UVGlobal 1000),
     (UVGlobal 1003, UVGlobal 1004),
     (UVGlobal 1004, UVGlobal 1003),
     (UVGlobal 1006, UVGlobal 1007),
     (UVGlobal 1007, UVGlobal 1006),
     (UVGlobal 1009, UVGlobal 1010),
     (UVGlobal 1010, UVGlobal 1009),
     (UVGlobal 1012, UVGlobal 1013),
     (UVGlobal 1013, UVGlobal 1012),
     (UVGlobal 1015, UVGlobal 1016),
     (UVGlobal 1016, UVGlobal 1015),
     (UVGlobal 1018, UVGlobal 1019),
     (UVGlobal 1019, UVGlobal 1018),
     (UVGlobal 1021, UVGlobal 1022),
     (UVGlobal 1022, UVGlobal 1021),
     (UVGlobal 1024, UVGlobal 1025),
     (UVGlobal 1025, UVGlobal 1024),
     (UVGlobal 1027, UVGlobal 1028),
     (UVGlobal 1028, UVGlobal 1027),
     (UVGlobal 1030, UVGlobal 1031),
     (UVGlobal 1031, UVGlobal 1030),
     (UVGlobal 1033, UVGlobal 1034),
     (UVGlobal 1034, UVGlobal 1033),
     (UVGlobal 1036, UVGlobal 1037),
     (UVGlobal 1037, UVGlobal 1036),
     (UVGlobal 1039, UVGlobal 1040),
     (UVGlobal 1040, UVGlobal 1039),
     (UVGlobal 1042, UVGlobal 1043),
     (UVGlobal 1043, UVGlobal 1042),
     (UVGlobal 1045, UVGlobal 1046),
     (UVGlobal 1046, UVGlobal 1045),
     (UVGlobal 1048, UVGlobal 1049),
     (UVGlobal 1049, UVGlobal 1048),
     (UVGlobal 1051, UVGlobal 1052),
     (UVGlobal 1052, UVGlobal 1051),
     (UVGlobal 1054, UVGlobal 1055),
     (UVGlobal 1055, UVGlobal 1054),
     (UVGlobal 1057, UVGlobal 1058),
     (UVGlobal 1058, UVGlobal 1057),
     (UVGlobal 1060, UVGlobal 1061),
     (UVGlobal 1061, UVGlobal 1060)],
  fromList []),
 Declare
   Meta
   (Rigid (NameIntro Meta (Id {unId = 0})))
   (MetaFunType
      Explicit
      (Rigid (TypeType Meta))
      (Rigid (TypeType Meta)))
   (Declare
      Meta
      (Rigid (NameIntro Meta (Id {unId = 6})))
      (Rigid (TypeType Meta))
      (Define
         (Rigid (NameIntro Meta (Id {unId = 0})))
         (Rigid (MetaConstIntro (Id {unId = 0})))
         (Define
            (Rigid (NameIntro Meta (Id {unId = 6})))
            (Rigid (MetaConstIntro (Id {unId = 6})))
            (Declare
               Meta
               (Rigid (NameIntro Meta (Id {unId = 1})))
               (MetaFunType
                  Explicit
                  (Rigid (TypeType Meta))
                  (MetaFunElim
                     (GlobalVar (Rigid (NameIntro Meta (Id {unId = 0}))))
                     (LocalVar (Index {unIndex = 0}))))
               (Declare
                  Meta
                  (Rigid (NameIntro Meta (Id {unId = 2})))
                  (MetaFunType
                     Explicit
                     (Rigid (TypeType Meta))
                     (MetaFunType
                        Explicit
                        (LocalVar (Index {unIndex = 0}))
                        (MetaFunType
                           Explicit
                           (MetaFunElim
                              (GlobalVar
                                 (Rigid (NameIntro Meta (Id {unId = 0}))))
                              (LocalVar (Index {unIndex = 1})))
                           (MetaFunElim
                              (GlobalVar
                                 (Rigid (NameIntro Meta (Id {unId = 0}))))
                              (LocalVar (Index {unIndex = 2}))))))
                  (Declare
                     Meta
                     (Rigid (NameIntro Meta (Id {unId = 3})))
                     (MetaFunType
                        Explicit
                        (Rigid (TypeType Meta))
                        (MetaFunType
                           Explicit
                           (MetaFunElim
                              (GlobalVar
                                 (Rigid (NameIntro Meta (Id {unId = 0}))))
                              (LocalVar (Index {unIndex = 0})))
                           (MetaFunType
                              Explicit
                              (MetaFunElim
                                 (GlobalVar
                                    (Rigid (NameIntro Meta (Id {unId = 0}))))
                                 (LocalVar (Index {unIndex = 1})))
                              (MetaFunType
                                 Explicit
                                 (MetaFunElim
                                    (GlobalVar
                                       (Rigid (NameIntro Meta (Id {unId = 0}))))
                                    (LocalVar (Index {unIndex = 2})))
                                 (Rigid (TypeType Meta))))))
                     (Declare
                        Meta
                        (Rigid (NameIntro Meta (Id {unId = 7})))
                        (GlobalVar (Rigid (NameIntro Meta (Id {unId = 6}))))
                        (Declare
                           Meta
                           (Rigid (NameIntro Meta (Id {unId = 8})))
                           (MetaFunElim
                              (GlobalVar
                                 (Rigid (NameIntro Meta (Id {unId = 0}))))
                              (GlobalVar
                                 (Rigid (NameIntro Meta (Id {unId = 6})))))
                           (Define
                              (Rigid (NameIntro Meta (Id {unId = 1})))
                              (Rigid (MetaConstIntro (Id {unId = 1})))
                              (Define
                                 (Rigid (NameIntro Meta (Id {unId = 2})))
                                 (Rigid (MetaConstIntro (Id {unId = 2})))
                                 (Define
                                    (Rigid (NameIntro Meta (Id {unId = 3})))
                                    (Rigid (MetaConstIntro (Id {unId = 3})))
                                    (Define
                                       (Rigid (NameIntro Meta (Id {unId = 7})))
                                       (Rigid (MetaConstIntro (Id {unId = 7})))
                                       (Define
                                          (Rigid
                                             (NameIntro Meta (Id {unId = 8})))
                                          (UniVar (UVGlobal 8))
                                          (Declare
                                             Meta
                                             (Rigid
                                                (NameIntro
                                                   Meta (Id {unId = 4})))
                                             (Rigid
                                                (AllType
                                                   (MetaFunIntro
                                                      (Rigid
                                                         (AllType
                                                            (MetaFunIntro
                                                               (MetaFunElim
                                                                  (MetaFunElim
                                                                     (MetaFunElim
                                                                        (MetaFunElim
                                                                           (GlobalVar
                                                                              (Rigid
                                                                                 (NameIntro
                                                                                    Meta
                                                                                    (Id
                                                                                       {unId =
                                                                                          3}))))
                                                                           (LocalVar
                                                                              (Index
                                                                                 {unIndex =
                                                                                    1})))
                                                                        (LocalVar
                                                                           (Index
                                                                              {unIndex =
                                                                                 0})))
                                                                     (MetaFunElim
                                                                        (GlobalVar
                                                                           (Rigid
                                                                              (NameIntro
                                                                                 Meta
                                                                                 (Id
                                                                                    {unId =
                                                                                       1}))))
                                                                        (LocalVar
                                                                           (Index
                                                                              {unIndex =
                                                                                 1}))))
                                                                  (MetaFunElim
                                                                     (GlobalVar
                                                                        (Rigid
                                                                           (NameIntro
                                                                              Meta
                                                                              (Id
                                                                                 {unId =
                                                                                    1}))))
                                                                     (LocalVar
                                                                        (Index
                                                                           {unIndex =
                                                                              1}))))))))))
                                             (Declare
                                                Meta
                                                (Rigid
                                                   (NameIntro
                                                      Meta (Id {unId = 5})))
                                                (Rigid
                                                   (AllType
                                                      (MetaFunIntro
                                                         (Rigid
                                                            (AllType
                                                               (MetaFunIntro
                                                                  (Rigid
                                                                     (AllType
                                                                        (MetaFunIntro
                                                                           (Rigid
                                                                              (AllType
                                                                                 (MetaFunIntro
                                                                                    (Rigid
                                                                                       (AllType
                                                                                          (MetaFunIntro
                                                                                             (Rigid
                                                                                                (ImplType
                                                                                                   (MetaFunElim
                                                                                                      (MetaFunElim
                                                                                                         (MetaFunElim
                                                                                                            (MetaFunElim
                                                                                                               (GlobalVar
                                                                                                                  (Rigid
                                                                                                                     (NameIntro
                                                                                                                        Meta
                                                                                                                        (Id
                                                                                                                           {unId =
                                                                                                                              3}))))
                                                                                                               (LocalVar
                                                                                                                  (Index
                                                                                                                     {unIndex =
                                                                                                                        4})))
                                                                                                            (LocalVar
                                                                                                               (Index
                                                                                                                  {unIndex =
                                                                                                                     2})))
                                                                                                         (LocalVar
                                                                                                            (Index
                                                                                                               {unIndex =
                                                                                                                  1})))
                                                                                                      (LocalVar
                                                                                                         (Index
                                                                                                            {unIndex =
                                                                                                               0})))
                                                                                                   (MetaFunElim
                                                                                                      (MetaFunElim
                                                                                                         (MetaFunElim
                                                                                                            (MetaFunElim
                                                                                                               (GlobalVar
                                                                                                                  (Rigid
                                                                                                                     (NameIntro
                                                                                                                        Meta
                                                                                                                        (Id
                                                                                                                           {unId =
                                                                                                                              3}))))
                                                                                                               (LocalVar
                                                                                                                  (Index
                                                                                                                     {unIndex =
                                                                                                                        4})))
                                                                                                            (MetaFunElim
                                                                                                               (MetaFunElim
                                                                                                                  (MetaFunElim
                                                                                                                     (GlobalVar
                                                                                                                        (Rigid
                                                                                                                           (NameIntro
                                                                                                                              Meta
                                                                                                                              (Id
                                                                                                                                 {unId =
                                                                                                                                    2}))))
                                                                                                                     (LocalVar
                                                                                                                        (Index
                                                                                                                           {unIndex =
                                                                                                                              4})))
                                                                                                                  (LocalVar
                                                                                                                     (Index
                                                                                                                        {unIndex =
                                                                                                                           3})))
                                                                                                               (LocalVar
                                                                                                                  (Index
                                                                                                                     {unIndex =
                                                                                                                        2}))))
                                                                                                         (LocalVar
                                                                                                            (Index
                                                                                                               {unIndex =
                                                                                                                  1})))
                                                                                                      (MetaFunElim
                                                                                                         (MetaFunElim
                                                                                                            (MetaFunElim
                                                                                                               (GlobalVar
                                                                                                                  (Rigid
                                                                                                                     (NameIntro
                                                                                                                        Meta
                                                                                                                        (Id
                                                                                                                           {unId =
                                                                                                                              2}))))
                                                                                                               (LocalVar
                                                                                                                  (Index
                                                                                                                     {unIndex =
                                                                                                                        4})))
                                                                                                            (LocalVar
                                                                                                               (Index
                                                                                                                  {unIndex =
                                                                                                                     3})))
                                                                                                         (LocalVar
                                                                                                            (Index
                                                                                                               {unIndex =
                                                                                                                  0})))))))))))))))))))))
                                                (Declare
                                                   Meta
                                                   (Rigid
                                                      (NameIntro
                                                         Meta (Id {unId = 9})))
                                                   (MetaFunElim
                                                      (MetaFunElim
                                                         (MetaFunElim
                                                            (MetaFunElim
                                                               (GlobalVar
                                                                  (Rigid
                                                                     (NameIntro
                                                                        Meta
                                                                        (Id
                                                                           {unId =
                                                                              3}))))
                                                               (GlobalVar
                                                                  (Rigid
                                                                     (NameIntro
                                                                        Meta
                                                                        (Id
                                                                           {unId =
                                                                              6})))))
                                                            (MetaFunElim
                                                               (GlobalVar
                                                                  (Rigid
                                                                     (NameIntro
                                                                        Meta
                                                                        (Id
                                                                           {unId =
                                                                              1}))))
                                                               (GlobalVar
                                                                  (Rigid
                                                                     (NameIntro
                                                                        Meta
                                                                        (Id
                                                                           {unId =
                                                                              6}))))))
                                                         (MetaFunElim
                                                            (GlobalVar
                                                               (Rigid
                                                                  (NameIntro
                                                                     Meta
                                                                     (Id
                                                                        {unId =
                                                                           1}))))
                                                            (GlobalVar
                                                               (Rigid
                                                                  (NameIntro
                                                                     Meta
                                                                     (Id
                                                                        {unId =
                                                                           6}))))))
                                                      (GlobalVar
                                                         (Rigid
                                                            (NameIntro
                                                               Meta
                                                               (Id
                                                                  {unId =
                                                                     8})))))
                                                   (Define
                                                      (Rigid
                                                         (NameIntro
                                                            Meta
                                                            (Id {unId = 4})))
                                                      (Rigid
                                                         (MetaConstIntro
                                                            (Id {unId = 4})))
                                                      (Define
                                                         (Rigid
                                                            (NameIntro
                                                               Meta
                                                               (Id {unId = 5})))
                                                         (Rigid
                                                            (MetaConstIntro
                                                               (Id {unId = 5})))
                                                         (Define
                                                            (Rigid
                                                               (NameIntro
                                                                  Meta
                                                                  (Id
                                                                     {unId =
                                                                        9})))
                                                            (Rigid
                                                               (MetaConstIntro
                                                                  (Id
                                                                     {unId =
                                                                        9})))
                                                            (Rigid
                                                               (TypeType
                                                                  Obj))))))))))))))))))))))
